src/lib.rs

```diff
@@ -1,182 +1,107 @@
-}
-
-/// Retrieves the name of a interface address
-unsafe fn get_ifa_name<'a>(ifa: *mut *mut ifaddrs) -> Result<&'a str, Error> {
-    CStr::from_ptr((**ifa).ifa_name)
-        .to_str()
-        .map_err(Error::IntAddrNameParseError)
-}
+//! `local-ip-address` is a wrapper on `getifaddrs` which retrieves host's
+//! network interfaces.
+//!
+//! Handy functions are provided such as `local_ip` which retrieve the local IP
+//! address based on the host system
+//!
+//! ```
+//! use std::net::IpAddr;
+//! use local_ip_address::local_ip;
+//!
+//! assert!(matches!(local_ip().unwrap(), IpAddr));
+//! ```
+//!
+//! You are able to iterate over a vector of tuples where the first element of
+//! the tuple is the name of the network interface and the second is the IP
+//! address.
+//!
+//! ```
+//! use std::net::IpAddr;
+//! use local_ip_address::find_af_inet;
+//!
+//! let ifas = find_af_inet().unwrap();
+//!
+//! if let Some((_, ipaddr)) = ifas
+//! .iter()
+//! .find(|(name, ipaddr)| *name == "en0" && matches!(ipaddr, IpAddr::V4(_))) {
+//!     println!("This is your local IP address: {:?}", ipaddr);
+//!     // This is your local IP address: 192.168.1.111
+//!     assert!(matches!(ipaddr, IpAddr));
+//! }
+//! ```
+//!
+
+use std::string::FromUtf8Error;
+use std::{env, net::IpAddr};
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    /// An error occured building a `&str` from a C string when
+    /// parsing the name of a interface address instance
+    #[error("Failed to read interface address name. `{0}`")]
+    IntAddrNameParseError(FromUtf8Error),
+    /// An error ocurred calling `getifaddrs`
+    #[error("Execution of getifaddrs had error result. getifaddrs returned `{0}`")]
+    GetIfAddrsError(i32),
+    /// The current platform is not supported
+    #[error("The current platform `{0}` is not supported")]
+    PlatformNotSupported(String),
+    #[error("GetIpAddrTableError")]
+    GetAdaptersAddresses(u32),
+}
+
+#[cfg(target_family = "unix")]
+pub mod unix;
+#[cfg(target_family = "unix")]
+use crate::unix::*;
+#[cfg(target_family = "windows")]
+pub mod windows;
+#[cfg(target_family = "windows")]
+use crate::windows::*;
+
+/// Perform a search over the system's network interfaces using `getifaddrs`,
+/// retrieved network interfaces belonging to both socket address families
+/// `AF_INET` and `AF_INET6` are retrieved along with the interface address name.
+///
+/// # Example
+///
+/// ```
+/// use std::net::IpAddr;
+/// use local_ip_address::find_af_inet;
+///
+/// let ifas = find_af_inet().unwrap();
+///
+/// if let Some((_, ipaddr)) = ifas
+/// .iter()
+/// .find(|(name, ipaddr)| *name == "en0" && matches!(ipaddr, IpAddr::V4(_))) {
+///     // This is your local IP address: 192.168.1.111
+///     println!("This is your local IP address: {:?}", ipaddr);
+/// }
+/// ```
+pub fn find_af_inet() -> Result<Vec<(String, IpAddr)>, Error> {
+    impl_find_af_inet()
+}

```

Comment:
We could avoid this if https://github.com/EstebanBorai/local-ip-address/commit/19c9e9fcae09cdea7200e045a1a93a4f1d74e3ca works as well AFAIK? Perhaps I'm missing something?